synapse/storage/data_stores/main/events_worker.py

```diff
@@ -1362,6 +1362,62 @@ def get_next_event_to_expire_txn(txn):
             desc="get_next_event_to_expire", func=get_next_event_to_expire_txn
         )
 
+    async def get_unread_message_count_for_user(
+        self, user_id: str, room_id: str, last_read_event_id: str,
+    ):
+        return await self.db.runInteraction(
+            "get_unread_message_count_for_user",
+            self._get_unread_message_count_for_user_txn,
+            user_id,
+            room_id,
+            last_read_event_id,
+        )
+
+    def _get_unread_message_count_for_user_txn(
+        self,
+        txn: LoggingTransaction,
+        user_id: str,
+        room_id: str,
+        last_read_event_id: Optional[str],
+    ):
+        if last_read_event_id:
+            # Get the stream ordering for the last read event.
+            stream_ordering = self.db.simple_select_one_onecol_txn(
+                txn=txn,
+                table="events",
+                keyvalues={"room_id": room_id, "event_id": last_read_event_id},
+                retcol="stream_ordering",
+            )
+        else:
+            # If there's no read receipt for that room, it probably means the user hasn't
+            # opened it yet, in which case use the stream ID of their join event.
+            # We can't just set it to 0 otherwise messages from other local users from
+            # before this user joined will be counted as well.
+            txn.execute(
+                """
+                SELECT stream_ordering FROM room_memberships
```

Comment:
room_memberships contains all room membership events *ever*: if the user comes and goes a few times, there is no guarantee you'll get the right one.

I think you want `local_current_membership`, but check that it's not background-populated or something annoying.

Reply:
Hmm right - I had updated `_maybe_insert_unread_event_txn` to use `current_state_events`, but apparently forgot that one.